src/main/groovy/pl/allegro/tech/build/axion/release/infrastructure/git/ScmCache.java

```diff
@@ -0,0 +1,106 @@
+package pl.allegro.tech.build.axion.release.infrastructure.git;
+
+import groovy.lang.Tuple2;
+import org.eclipse.jgit.api.Git;
+import org.eclipse.jgit.api.errors.GitAPIException;
+import org.eclipse.jgit.lib.Ref;
+import org.eclipse.jgit.revwalk.RevCommit;
+import org.eclipse.jgit.revwalk.RevWalk;
+import pl.allegro.tech.build.axion.release.domain.scm.ScmException;
+import pl.allegro.tech.build.axion.release.domain.scm.ScmRepository;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Provides cached version for some operations on {@link ScmRepository}
+ */
+public class ScmCache {
+
+    /**
+     * Since this cache is statis and Gradle Demon might keep JVM process in background for a long
+     * time, we have to put some TTL for cached values.
+     */
+    private static final long INVALIDATE_AFTER_MILLIS = 1000 * 60;
+
+    private static final ScmCache CACHE = new ScmCache();
+
+    public static ScmCache getInstance() {
+        return CACHE;
+    }
+
+    private ScmCache() { }
+
+    private final Map<String, CachedState> cache = new HashMap<>();
+
+    synchronized void invalidate(ScmRepository repository) {
+        cache.remove(repository.id());
+    }
+
+    public synchronized boolean checkUncommittedChanges(ScmRepository repository) {
```

Comment:
wouldn't a concurrent hash map be better than a synchronized block?

Reply:
possibly. do you think it will be noticeable improvement?